linktimeplugin.hpp

```diff
@@ -59,32 +53,33 @@ namespace linktimeplugin {
         // Implemented by the derived registrar class.
         virtual BASE& operator()() = 0;
 
+        typedef std::vector<BASE*> PluginColl;
+        typedef std::vector<RegistrarBase<BASE>*> RegistarColl;
+        
         // Returns all registrars.
-        static std::vector<BASE*> plugins() {
-            std::vector<BASE*> ret;
-
-            if (registrars_) {
-                for(auto r : *registrars_) {
-                    ret.push_back(&(*r)());
-                }
+        static PluginColl plugins() {
+            PluginColl ret;
+            for(auto r : registrars()) {
+                ret.push_back(&(*r)());
             }
-
             return ret;
         }
-
+        // Use (*begin())() to get BASE object!
+        static auto begin() {
+            return registrars().begin();
+        }
+        
+        static auto end() {
+            return registrars().end();
+        }
+        
     private:
-        // Pointers to the registrar objects (one per registered
-        // plug-in class).
-        static std::unique_ptr<std::vector<RegistrarBase<BASE>*>>registrars_;
+        static RegistarColl& registrars() {
+            static RegistarColl regs;
```

Comment:
This calls the std::vector ctor which may throw, which would make the RegistrarBase ctor violate its noexcept promise. I think, however, that the static object hidden in a private function is superior to the smart pointer used previously.

Reply:
My bad. I only considered the allocation in the original code and not push_back which obviously can trigger reallocation which can throw bad_alloc. try-catch is needed then.